### PR TITLE
networkd: ListLLDPNeighbours bus method

### PR DESCRIPTION
**Pull request by peckato1**
Systemd (networkd) does not yet allow programatically query LLDP neighbours (not counting parsing of `networkctl lldp` output) from the outside world. This patch allows listing LLDP neighbours via a DBus method call. 

-----

Tbh I am not sure whether it is better to list all neighbours for all links (using networkd-manager bus method as in this patch) or create a method on networkd-link object that lists only neighbours for the specified interface or do both ways.

### CODE REVIEW

**Comment by poettering** [comment]
> Tbh I am not sure whether it is better to list all neighbours for all links (using networkd-manager bus method as in this patch) or create a method on networkd-link object that lists only neighbours for the specified interface or do both ways.

we often have both actually, since dbus roundtrips are expensive and if there's a common case to have something on the manager object then we can add it.

**Comment by poettering** [comment]
i am a bit concerned about sizes here... the number of lldp neighbours might be huge, and dbus isn't really designed for passing around such effectively unbounded things.

the lldp peers can be parsed from /run/systemd/netif/lldp/. maybe we should make that official and make sd-lldp.h public so that clients can parse these things client side.

**Comment by peckato1** [comment]
> the number of lldp neighbours might be huge, and dbus isn't really designed for passing around such effectively unbounded things.

I did not realize that dbus might not be the best for that :-)

> the lldp peers can be parsed from /run/systemd/netif/lldp/.
> maybe we should make that official and make sd-lldp.h public so that clients can parse these things client side.

Yes, that where `networkctl lldp` command takes the data from, I noticed that. However, we were not convinced that parsing the binary data is a good approach. That's why I read the peers from internal data structures.

**Comment by peckato1** [comment]
Hi, 

so is there some update on what variant would you prefer?

Best,
Tomas

**Comment by peckato1** [comment]
I tried to make sd_lldp public in the latest version as proposed. Is this approach be acceptable?

**Comment by bluca** [comment]
Needs a rebase

File: src/fuzz/meson.build [hunk 1]
```diff
@@ -33,8 +33,8 @@ fuzzers += [
          []],
 
         [['src/fuzz/fuzz-lldp.c'],
-         [libshared,
-          libsystemd_network],
+         [libbasic,libshared_static,libsystemd_static
+          ],
          []],
```
Comment (keszybz) [inline src/fuzz/meson.build]: Please follow the existing style: one name per line.

File: src/fuzz/meson.build [hunk 2]
```diff
@@ -43,8 +43,9 @@ fuzzers += [
           'src/libsystemd-network/icmp6-util.h',
           'src/systemd/sd-dhcp6-client.h',
           'src/systemd/sd-ndisc.h'],
-         [libshared,
-          libsystemd_network],
+         [libsystemd_network,
+          libshared
+          ],
```
Comment (keszybz) [inline src/fuzz/meson.build]: Why this change?

File: src/libsystemd/meson.build [hunk 3]
```diff
@@ -70,6 +70,12 @@ libsystemd_sources = files('''
         sd-hwdb/hwdb-util.c
         sd-hwdb/hwdb-util.h
         sd-hwdb/sd-hwdb.c
+        sd-lldp/sd-lldp.c
+        sd-lldp/lldp-internal.h
+        sd-lldp/lldp-neighbor.c
+        sd-lldp/lldp-neighbor.h
+        sd-lldp/lldp-network.c
+        sd-lldp/lldp-network.h
```
Comment (keszybz) [inline src/libsystemd/meson.build:78]: I don't think we want to move `lldp` to be part of systemd. I know that system-network is not exposed as a public library. Most likely we should start doing that.

File: src/test/meson.build [hunk 4]
```diff
@@ -1042,6 +1042,11 @@ tests += [
          [],
          []],
 
+        [['src/libsystemd/sd-lldp/test-lldp.c'],
+         [libbasic,libshared_static,libsystemd_static
+          ],
```
Comment (keszybz) [inline src/test/meson.build]: Please folllow the usual style: one per line, no concatenation without whitespace.

**Review by keszybz — COMMENTED** [review]
This PR moves parts of libsystemd-network to libsystemd to make them public. This doesn't seem to be the right approach. We should either make libsystemd-network public, or, if that is too big of a step, we should make libsystemd-lldp public as a separate library.

**Comment by peckato1** [comment]
> This PR moves parts of libsystemd-network to libsystemd to make them public. This doesn't seem to be the right approach. 

Hello, thanks for your input and sorry for my delayed response. I agree with you.

> We should either make libsystemd-network public, or, if that is too big of a step, we should make libsystemd-lldp public as a separate library.

What is your preference here? I am willing to help but that sounds more like a design decision for maintainers.

**Comment by peckato1** [comment]
I tried to create the systemd-lldp public library you proposed. So perhaps we can start from there? See https://github.com/peckato1/systemd/commit/3f1a36b9b9fd7a5c12a1ea3a443c92df78fc86e7

**Comment by yuwata** [comment]
Hmm, I suggest that networkctl should allow to output LLDP in json format. Then, people can easily use it. I do not think sd-lldp is solid well to be public yet.

**Comment by peckato1** [comment]
Hello,
 I'm willing to help in implementing that as well. However, I'd still like to know what is the preferred way because now it seems there are several opinions :-)

Regards,
Tomas

**Comment by poettering** [comment]
Let's continue this in  #20333
